src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt

```diff
@@ -0,0 +1,634 @@
+        element: PsiElement
+    ) : RsDiagnostic(element) {
+        override fun addToHolder(holder: AnnotationHolder): Annotation {
+            return holder.createErrorAnnotation(
+                element,
+                header(RsErrorCode.E0379, "Trait functions cannot be declared const"),
+                ""
+            )
+        }
+    }
+}
+
+sealed class RsTextRangeDiagnostic(val element: TextRange) {
+    abstract fun addToHolder(holder: AnnotationHolder): Annotation
+
+    class TraitItemsMissingImplError(
+        element: TextRange,
+        private val missing: String
+    ) : RsTextRangeDiagnostic(element) {
+        override fun addToHolder(holder: AnnotationHolder): Annotation {
+            return holder.createErrorAnnotation(
+                element,
+                header(RsErrorCode.E0046, errorText()),
+                ""
+            )
+        }
+
+        private fun errorText(): String {
+            val missingS = escapeString(missing)
+            return "Not all trait items implemented, missing: $missingS"
+        }
+    }
+
+    class CrateNotFoundError(
+        element: TextRange,
+        private val crateName: String
+    ) : RsTextRangeDiagnostic(element) {
+        override fun addToHolder(holder: AnnotationHolder): Annotation {
+            return holder.createErrorAnnotation(
+                element,
+                header(RsErrorCode.E0463, errorText()),
+                ""
+            )
+        }
+
+        private fun errorText(): String {
+            val elTextS = escapeString(crateName)
+            return "Can't find crate for `$elTextS`"
+        }
+    }
+}
+
+sealed class RsRequireResolveDiagnostic(val element: RsReferenceElement) {
+    abstract fun requireResolve(holder: AnnotationHolder)
+
+    class UndeclaredLabelError(
+        element: RsReferenceElement
+    ): RsRequireResolveDiagnostic(element) {
+        override fun requireResolve(holder: AnnotationHolder) {
+            holder.requireResolve(
+                element,
+                header(RsErrorCode.E0426, errorText()),
+                ""
+            )
+        }
+
+        private fun errorText(): String {
+            val labelText = escapeString(element.text)
+            return "Use of undeclared label `$labelText`"
+        }
+    }
+
+    class UndeclaredLifetimeError(
+        element: RsReferenceElement
+    ): RsRequireResolveDiagnostic(element) {
+        override fun requireResolve(holder: AnnotationHolder) {
+            holder.requireResolve(
+                element,
+                header(RsErrorCode.E0261, errorText()),
+                ""
+            )
+        }
+
+        private fun errorText(): String {
+            val lifetimeText = escapeString(element.text)
+            return "Use of undeclared lifetime name `$lifetimeText`"
+        }
+    }
+}
+
+private class AnnotationHeader(
+    val plainText: String,
+    val html: String
+)
+
+private enum class RsErrorCode {
+    E0046,
```

Comment:
I need to check kotlin style guide, but I think it's ok to write a multiple declarations on a single line

Reply:
Hmm, no, I decided to leave it as is =D. In future we will insert more codes into middle of the list, so any complex formatting will be hard to maintain...